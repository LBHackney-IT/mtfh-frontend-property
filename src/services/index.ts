export * from "./config";
export { default as locale } from "./locale";
