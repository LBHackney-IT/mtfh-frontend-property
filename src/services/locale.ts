import { AssetAddress, AssetType } from "@mtfh/common/lib/api/asset/v1";

export default {
  backToSearch: "Search results",
  cancel: "Cancel",
  noDataForKey: "N/A",
  assets: {
    assetDetails: {
      address: (assetAddress: AssetAddress): string => {
        const {
          postPreamble,
          addressLine1,
          addressLine2,
          addressLine3,
          addressLine4,
          postCode,
        } = assetAddress;
  
        return [
          postPreamble,
          addressLine1,
          addressLine2,
          addressLine3,
          addressLine4,
          postCode,
        ]
          .filter((addressLine) => !!addressLine)
          .join(" ");
      },
      newTenure: "New tenure",
    },
    assetType: (assetType: AssetType): string => {
      if (assetType === "Dwelling") {
        return "Dwelling";
      }
      if (assetType === "LettableNonDwelling") {
        return "Lettable non-dwelling";
      }
      return assetType;
    },
    patchAssetAddressSuccessMessage: "The asset address has been updated successfully."
  },
  comments: {
    heading: "Comments",
    addComment: "Add comment",
  },
  repairs: {
    heading: "Repairs",
  },
  tenureDetails: {
    tenureLabel: "Tenure",
    expandedTenureSection: "Tenure details",
    viewTenureButtonLabel: "View tenure",
    assetTypeLabel: "Type",
    assetPaymentReferenceLabel: "Payment reference",
    uprnLabel: "UPRN",
    assetReferenceLabel: "Reference",
    active: "Active",
    isActive: (isActive: boolean): string => (isActive ? "Active" : "Inactive"),
    status: "Status",
    type: "Type",
    startDate: "Start date",
    endDate: "End date",
    noTenure: "No tenure",
    paymentRef: "Payment ref.",
  },
  cautionaryAlerts: {
    cautionaryAlerts: "Cautionary alerts",
    none: "None",
  },
  static: {
    newProcess: "New Process",
  },
  assetCouldNotBeLoaded: "This property could not be loaded.",
  errors: {
    unableToFetchRecord: "There was a problem retrieving the record",
    assetDoesNotExist: "The property information you've requested does not exist",
    unableToFetchRecordDescription:
      "Please try again. If the issue persists, please contact support.",
    alertTitle: "Error",
    postcodeInvalid: "Please enter a valid postcode",
    unableToPatchAsset: "There was a problem amending this asset",
    unableToPatchAssetInvalidVersion: "The 'version' property is invalid on this asset. This is a required property when updating the asset.",
    unableToPatchAssetDescription:
      "Please refresh the page and try again. If the issue persists, please contact support.",
  },
};
