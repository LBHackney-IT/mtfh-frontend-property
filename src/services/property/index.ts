export * from './property';
export * from './property.types';
