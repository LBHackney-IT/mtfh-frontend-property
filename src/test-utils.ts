import {
    getAssetV1,
    server,
    getCommentV2,
    getReferenceDataV1,
    mockCommentReferenceDataV1,
} from '@hackney/mtfh-test-utils';

beforeEach(() => {
    server.use(getAssetV1());
    server.use(getCommentV2());
    server.use(getReferenceDataV1([...mockCommentReferenceDataV1]));
});
