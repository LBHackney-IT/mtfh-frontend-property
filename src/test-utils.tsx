import { SWRConfig } from 'swr';
import { Route, Router } from 'react-router-dom';
import React from 'react';
import MatchMediaMock from 'jest-matchmedia-mock';
import { createMemoryHistory } from 'history';
import { RenderResult, render } from '@testing-library/react';
import { queries } from '@mtfh/common';

let matchMedia: MatchMediaMock;

beforeAll(() => {
    matchMedia = new MatchMediaMock();
});

interface RouteRenderConfig {
    url: string;
    path: string;
    query: keyof typeof queries;
}

export const routeRender = (
    component: JSX.Element,
    options?: Partial<RouteRenderConfig>
): RenderResult => {
    const config: RouteRenderConfig = {
        url: '/tenure/68c6896c-16f1-54d2-3504-847cb438a1b1',
        path: '/tenure/:tenureId',
        query: 'lg',
        ...options,
    };
    matchMedia.useMediaQuery(`(min-width: 0px)`);
    const history = createMemoryHistory();
    history.push(config.url);
    matchMedia.useMediaQuery(queries[config.query]);
    history.push(config.url);
    return render(
        <SWRConfig value={{ dedupingInterval: 0, errorRetryInterval: 0 }}>
            <Router history={history}>
                <Route path={config.path}>{component}</Route>
            </Router>
        </SWRConfig>
    );
};
