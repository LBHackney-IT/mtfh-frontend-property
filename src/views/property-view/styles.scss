@import '~lbh-frontend/lbh/base';

.mtfh-property-sidebar {
    .lbh-button {
        width: 100%;
    }
}
