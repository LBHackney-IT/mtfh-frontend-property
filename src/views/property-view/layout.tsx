import { Link as RouterLink } from 'react-router-dom';
import React, { FC } from 'react';

import {
    hasToggle,
    Button,
    Layout,
    Link,
    PageAnnouncementProvider,
    PageAnnouncement,
    SideBar,
    SideBarSection,
    SideBarProps,
} from '@mtfh/common';
import { isFutureDate } from '../../utils';
import { locale, Property } from '../../services';
import { PropertyDetails, TenureDetails } from '../../components';
import './styles.scss';

export interface PropertyLayoutProperties {
    propertyDetails: Property;
}

interface PropertySideBarProperties
    extends Partial<SideBarProps>,
        PropertyLayoutProperties {}

const PropertySideBar = ({
    propertyDetails,
    ...properties
}: PropertySideBarProperties) => {
    const { assetAddress, assetId, assetType, tenure, id } = propertyDetails;

    return (
        <div className="mtfh-property-sidebar">
            <SideBar id="property-view-sidebar" {...properties}>
                <PropertyDetails
                    assetAddress={assetAddress}
                    assetType={assetType}
                    propertyReference={assetId}
                />
                <SideBarSection
                    id="tenure-details"
                    title={locale.tenureDetails.expandedTenureSection}
                >
                    <TenureDetails tenure={tenure} />
                </SideBarSection>
            </SideBar>
            {hasToggle('MMH.CreateTenure') &&
                (!tenure ||
                    !tenure.isActive ||
                    !isFutureDate(tenure.endOfTenureDate)) && (
                    <Button as={RouterLink} to={`/tenure/${id}/add`}>
                        {locale.propertyDetails.newTenure}
                    </Button>
                )}
        </div>
    );
};

export const PropertyLayout: FC<PropertyLayoutProperties> = ({
    propertyDetails,
}) => {
    return (
        <PageAnnouncementProvider sessionKey="property">
            <PageAnnouncement />
            <Layout
                top={
                    <>
                        <Link as={RouterLink} to="/search" variant="back-link">
                            {locale.backToSearch}
                        </Link>
                        <h1 className="heading">
                            {locale.propertyDetails.address(
                                propertyDetails.assetAddress
                            )}
                        </h1>
                    </>
                }
                side={<PropertySideBar propertyDetails={propertyDetails} />}
            />
        </PageAnnouncementProvider>
    );
};
