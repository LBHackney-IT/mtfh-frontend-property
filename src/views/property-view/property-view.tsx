import { useParams } from 'react-router-dom';
import React from 'react';
import { Center, ErrorSummary, Spinner } from '@mtfh/common';
import { PropertyLayout } from './layout';
import { locale, usePropertyMock } from '../../services';
export const PropertyView = (): JSX.Element => {
    const { propertyId } = useParams<{ propertyId: string }>();

    const { data: property, error } = usePropertyMock(propertyId);

    if (error) {
        return (
            <ErrorSummary
                id="property-error"
                title={locale.errors.unableToFetchRecord}
                description={locale.errors.unableToFetchRecordDescription}
            />
        );
    }

    if (!property) {
        return (
            <Center>
                <Spinner />
            </Center>
        );
    }

    return <PropertyLayout propertyDetails={property} />;
};
