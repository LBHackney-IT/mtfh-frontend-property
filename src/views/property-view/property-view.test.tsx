import React from 'react';
import { screen, waitFor } from '@testing-library/react';

import { featureToggleStore } from '@mtfh/common';
import { PropertyView } from '../property-view';
import { routeRender, get } from '../../test-utils';
import { locale } from '../../services';
import {
    mockProperty,
    mockPropertyLettableNonDwelling,
    mockPropertyInvalidAssetType,
} from '../../mocks/data';

test('renders the error on Property failure', async () => {
    get(`/api/assets/:id`, { message: 'failure' }, 500);
    routeRender(<PropertyView />, { url: `/assets/${mockProperty.id}` });

    await waitFor(() => expect(screen.getByRole('alert')).toBeInTheDocument());
});

test('renders the property view', async () => {
    const [{ container }] = routeRender(<PropertyView />);
    expect(container).toMatchSnapshot();

    await waitFor(() =>
        expect(screen.getAllByRole('heading')[0]).toHaveTextContent(
            locale.propertyDetails.address(mockProperty.assetAddress)
        )
    );
    expect(screen.queryByText(/Dwelling/)).toBeInTheDocument();

    expect(screen.queryByText(/UPRN/)).toBeInTheDocument();
    expect(screen.queryByText(/100021065786/)).toBeInTheDocument();

    expect(screen.queryByText(/Reference/)).toBeInTheDocument();
});

test('it shows the back button', async () => {
    routeRender(<PropertyView />);

    await waitFor(() =>
        expect(screen.getByText('Search results')).toBeInTheDocument()
    );
});

test('it shows new tenure button', async () => {
    const features = featureToggleStore.getValue();
    featureToggleStore.next({ MMH: { ...features.MMH, CreateTenure: true } });
    routeRender(<PropertyView />);

    await waitFor(() =>
        expect(
            screen.getByText(locale.propertyDetails.newTenure)
        ).toBeInTheDocument()
    );
});

test('renders the property view for lettable-non-dwelling', async () => {
    get(`/api/assets/:id`, mockPropertyLettableNonDwelling, 200);
    routeRender(<PropertyView />);

    await waitFor(() =>
        expect(screen.queryByText(/Lettable non-dwelling/)).toBeInTheDocument()
    );
});

test('renders the property view for invalid asset type', async () => {
    get(`/api/assets/:id`, mockPropertyInvalidAssetType, 200);
    routeRender(<PropertyView />);

    await waitFor(() =>
        expect(
            screen.queryByText(/This asset could not be loaded./)
        ).toBeInTheDocument()
    );
});
