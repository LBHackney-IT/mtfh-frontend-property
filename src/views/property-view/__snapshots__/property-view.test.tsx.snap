// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the property view 1`] = `
<div>
  <div
    class="mtfh-center mtfh-center--horizontal mtfh-center--vertical"
  >
    <svg
      class="mtfh-icon"
      focusable="false"
      height="50"
      stroke="#00703c"
      viewBox="0 0 42 42"
      width="50"
    >
      <title>
        Loading...
      </title>
      <g
        fill="none"
        fill-rule="evenodd"
      >
        <g
          stroke-width="5"
          transform="translate(3 3)"
        >
          <circle
            cx="18"
            cy="18"
            r="18"
            stroke-opacity=".5"
          />
          <path
            d="M36 18c0-9.94-8.06-18-18-18"
            transform="rotate(112.708 18 18)"
          >
            <animatetransform
              attributeName="transform"
              dur="1s"
              from="0 18 18"
              repeatCount="indefinite"
              to="360 18 18"
              type="rotate"
            />
          </path>
        </g>
      </g>
    </svg>
  </div>
</div>
`;
